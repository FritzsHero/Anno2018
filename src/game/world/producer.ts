/*
 * The savegame data structure was reverse-engineered by Benedikt Freisen
 * as part of the 'mdcii-engine' project and released under GPLv2+.
 * https://github.com/roybaer/mdcii-engine
 */

import { Point } from "pixi.js";
import Stream from "../../parsers/stream";
import assert from "../../util/assert";

export class Producer {
  public islandId: number;
  public position: Point;

  /**
   * The amount of goods already produced and waiting for pickup.
   */
  public stock: number;
  // The amount of primary input goods waiting to be used.
  // They are only "taken" when an output good is finished.
  public firstGoodStock: number;
  // The amount of secondary input goods waiting to be used (normally wood).
  // This is 0 for buildings which don't use secondary goods.
  // They are only "taken" when an output good is finished.
  public secondGoodStock: number;

  /**
   * The timer is decremented every second, regardless of whether or not
   * the building has enough input goods to produce something. Once it would reach 0
   * (i.e., the timer never actually reaches 0):
   * If there are enough input goods to produce an output good (= producedGood is 128):
   *   - the timer is set to "HAUS_PRODTYP.Interval"
   *   - "stock" is incremented
   *   - "firstGoodStock" is decremented by "HAUS_PRODTYP.Rohmenge"
   *   - "secondGoodStock" ???
   *   - "noGoodCnt" is set to 0
   * If there are NOT enough goods (= producedGood is not 128):
   *   - it is set to 11 (at least for "Webstube")
   *   - "noGoodCnt" is increased (up to a maximum of 15)
   *
   * Whenever new input goods are delivered and there were previously not enough
   * input goods to create output goods, the timer is immediately set to
   * "HAUS_PRODTYP.Interval" ??
   *
   * I am unsure why the timer is running even when no goods are currently being
   * produced.
   */
  public timer: number;
  public prodCount: number;
  public timeCount: number;
  public howMuchIsBeingProduced: number;

  private goodWasProduced: number;
  private speedCnt: number;

  public static load(data: Stream) {
    const p = new Producer();

    p.islandId = data.read8();
    p.position = new Point(data.read8(), data.read8());
    data.read8();

    p.speedCnt = data.read8() & 0b111;
    p.stock = data.read16();
    data.read8();

    p.timer = data.read16();
    p.secondGoodStock = data.read16();

    p.firstGoodStock = data.read16();
    data.read8();
    p.howMuchIsBeingProduced = data.read8();

    p.prodCount = data.read16();
    p.timeCount = data.read16();

    const flags = data.read8();
    p.setActive(!!(flags & (1 << 0)));
    p.setAnimCount((flags >> 2) & 0b1111);
    p.setGoodsAllowedForPickup((flags & (1 << 6)) === 1);
    if (flags & (1 << 1)) {
      p.goodWasProduced |= 1 << 6;
    } else {
      p.goodWasProduced &= ~(1 << 6);
    }

    p.setGoodWasProducedTimer(data.read8() & 0b1111);
    assert(data.read16() === 0);

    return p;
  }

  public save(stream: Stream) {
    stream.write8(this.islandId);
    stream.write8(this.position.x);
    stream.write8(this.position.y);
    stream.write8(0);

    stream.write8(this.speedCnt & 0b111);
    stream.write16(this.stock);
    stream.write8(0);

    stream.write16(this.timer);
    stream.write16(this.secondGoodStock);

    stream.write16(this.firstGoodStock);
    stream.write8(0);
    stream.write8(this.howMuchIsBeingProduced);

    stream.write16(this.prodCount);
    stream.write16(this.timeCount);

    const flags =
      ((this.isActive() ? 0b1 : 0b0) << 0) |
      (((this.goodWasProduced >> 6) & 1) << 1) |
      (this.getAnimCount() << 2) |
      ((this.goodsAllowedForPickup() ? 0b1 : 0b0) << 6);
    stream.write8(flags);
    stream.write8(this.getGoodWasProducedTimer());
    stream.write16(0);
  }

  public isProducing() {
    return !!(this.goodWasProduced & (1 << 7));
  }

  public isActive() {
    return !!(this.speedCnt & (1 << 3));
  }

  public setActive(active: boolean) {
    if (active) {
      this.speedCnt |= 1 << 3;
    } else {
      this.speedCnt &= ~(1 << 3);
    }
  }

  public goodsAllowedForPickup() {
    return !!(this.goodWasProduced & (1 << 4));
  }

  public setGoodsAllowedForPickup(value: boolean) {
    if (value) {
      this.goodWasProduced |= 1 << 4;
    } else {
      this.goodWasProduced &= ~(1 << 4);
    }
  }

  public stockAsInt() {
    return this.stock >>> 5;
  }

  public firstGoodStockAsInt() {
    return this.firstGoodStock >>> 5;
  }

  public secondGoodStockAsInt() {
    return this.secondGoodStock >>> 5;
  }

  /**
   * timer that counts up and is reset whenever a good is produced
   */
  public getGoodWasProducedTimer() {
    return this.goodWasProduced & 0b1111;
  }

  public setGoodWasProducedTimer(value: number) {
    assert(value >= 0 && value <= 0b1111);
    this.goodWasProduced =
      (this.goodWasProduced & 0b11110000) | (value & 0b1111);
  }

  public getAnimCount() {
    return (this.speedCnt >> 4) & 0b1111;
  }

  public setAnimCount(value: number) {
    assert(value >= 0 && value <= 0b1111);
    this.speedCnt = (this.speedCnt & 0b1111) | (value << 4);
  }
}
