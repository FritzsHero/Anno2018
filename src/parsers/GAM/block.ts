import Stream from "../stream";

export class Block {
    public static empty(type: string) {
        return new Block(type, 0, new Stream(new Uint8Array()));
    }

    public readonly data: Stream;

    constructor(
        public readonly type: string,
        public readonly length: number,
        data: Stream,
    ) {
        this.data = new Stream(data.slice(length));
    }
}

// tslint:disable-next-line:max-classes-per-file
export class IslandBlock extends Block {
    public inselHausBlocks: Block[] = [];
}
