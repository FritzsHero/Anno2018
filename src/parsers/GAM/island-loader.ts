import FileSystem from "../../filesystem";
import Island from "../../game/world/island";

export type IslandSizeId = 0|1|2|3|4;
export type IslandSizeName = "lit"|"mit"|"med"|"big"|"lar";
interface IslandSize {
    id: IslandSizeId;
    name: IslandSizeName;
    maxSize: number;
}

export default class IslandLoader {
    private readonly islandSizes: IslandSize[] = [
        {
            id: 0,
            name: "lit",
            maxSize: 35,
        },
        {
            id: 1,
            name: "mit",
            maxSize: 45,
        },
        {
            id: 2,
            name: "med",
            maxSize: 55,
        },
        {
            id: 3,
            name: "big",
            maxSize: 85,
        },
        {
            id: 4,
            name: "lar",
            maxSize: 100,
        },
    ];

    constructor(private fs: FileSystem) {}

    public async load(island: Island) {
        const climate = island.isSouth ? "south" : "north";
        const islandNumber = island.baseIslandNumber.toString().padStart(2, "0");
        for (const islandSize of this.islandSizes) {
            if (island.width <= islandSize.maxSize) {
                return await this.fs.openAndGetContentAsStream(
                    `/islands/${climate}/${islandSize.name}${islandNumber}.scp`,
                );
            }
        }

        throw new Error("Could not load island");
    }

    public async loadRandom(sizeId: 0|1|2|3|4, climate: "NORTH"|"SOUTH"|"ANY") {
        const islandSize = this.islandSizes.find((size) => size.id === sizeId);
        if (islandSize === undefined) {
            throw new Error("This should never happen.");
        }

        const islandFiles = [];
        if (["ANY", "NORTH"].includes(climate)) {
            islandFiles.push(...await this.loadIslandsWithSizeAndClimate(islandSize.name, "north"));
        }
        if (["ANY", "SOUTH"].includes(climate)) {
            islandFiles.push(...await this.loadIslandsWithSizeAndClimate(islandSize.name, "south"));
        }

        const randomIslandFile = islandFiles[Math.floor(Math.random() * islandFiles.length)];

        console.log(randomIslandFile.fullPath);

        return await this.fs.openAndGetContentAsStream(randomIslandFile.fullPath);
    }

    private async loadIslandsWithSizeAndClimate(sizeName: IslandSizeName, climate: "north"|"south") {
        return (await this.fs.ls(`/islands/${climate}`)).filter((islandFile) => islandFile.name.startsWith(sizeName));
    }
}
