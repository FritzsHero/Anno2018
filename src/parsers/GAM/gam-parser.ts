/*
 * The savegame data structure was reverse-engineered by Benedikt Freisen
 * as part of the 'mdcii-engine' project and released under GPLv2+.
 * https://github.com/roybaer/mdcii-engine
 */

import * as assert from "assert";
import Castle from "../../game/world/castle";
import City from "../../game/world/city";
import Field from "../../game/world/field";
import Island from "../../game/world/island";
import Kontor from "../../game/world/kontor";
import Player from "../../game/world/player";
import Ship from "../../game/world/ship";
import Soldier from "../../game/world/soldier";
import Task from "../../game/world/task";
import Trader from "../../game/world/trader";
import World from "../../game/world/world";
import Stream from "../stream";
import {Block, IslandBlock} from "./block";
import IslandLoader from "./island-loader";

export interface IslandField {
    fieldId: number;
    x: number;
    y: number;
    rotation: number;
    ani: number;
    _1: number;
    status: number;
    random: number;
    player: number;
    _2: number;
}

export type PlayerMap = ReadonlyMap<number, Player>;
export type IslandMap = ReadonlyMap<number, Island>;

interface Entity<T> {
    fromSaveGame(data: Stream, players: PlayerMap, islands: IslandMap): T;
}

export default class GAMParser {
    constructor(private islandLoader: IslandLoader) { }

    public async parse(data: Stream) {
        const blocks: Map<string, Block[]> = new Map();
        while (!data.eof()) {
            const block = this.readBlock(data);
            console.log(block.type);
            if (block.type !== "INSELHAUS") {
                if (!blocks.has(block.type)) {
                    blocks.set(block.type, []);
                }
                blocks.get(block.type).push(block);
            } else {
                const lastIslandBlock = blocks.get("INSEL5")[blocks.get("INSEL5").length - 1] as IslandBlock;
                if (lastIslandBlock.inselHausBlocks === undefined) {
                    lastIslandBlock.inselHausBlocks = [];
                }
                lastIslandBlock.inselHausBlocks.push(block);
            }
        }

        return this.doParse(blocks);
    }

    private async doParse(blocks: Map<string, Block[]>) {
        // console.log([...blocks.keys()].map(key => [key, blocks.get(key).length]));
        console.log(
            [...blocks.keys()]
                .filter((key) => blocks.get(key).find((block) => block.length > 0) !== undefined)
                .map((key) => [key, blocks.get(key)]),
        );

        let gameName = "";
        if (blocks.has("NAME")) {
            // Missions don't have a name.
            const nameBlock = blocks.get("NAME")[0];
            gameName = nameBlock.data.readString(nameBlock.length);
        }

        let task = null;
        if (blocks.has("AUFTRAG4")) {
            const taskBlock = blocks.get("AUFTRAG4")[0];
            if (taskBlock.length > 0) {
                task = Task.fromSaveGame(taskBlock.data);
            }
        }

        const playerBlock = blocks.get("PLAYER4")[0];
        const players = this.parsePlayers(playerBlock);

        const islandBlocks = blocks.has("INSEL5") ? blocks.get("INSEL5") as IslandBlock[] : [];
        const islands = await this.parseIslands(islandBlocks, players);

        const ships    = this.handleBlock<Ship>(   blocks, "SHIP4",   Ship,    players, islands);
        const soldiers = this.handleBlock<Soldier>(blocks, "SOLDAT3", Soldier, players, islands);
        const kontors  = this.handleBlock<Kontor>( blocks, "KONTOR2", Kontor,  players, islands);
        const castles  = this.handleBlock<Castle>( blocks, "MILITAR", Castle,  players, islands);
        const cities   = this.handleBlock<City>(   blocks, "STADT4",  City,    players, islands);

        // TODO: HIRSCH2, PRODLIST2, WERFT, SIEDLER, ROHWACHS2, MARKT2, HANDLER, TURM, TIMERS, WIFF
        let trader = null;
        if (blocks.has("HANDLER") && blocks.get("HANDLER")[0].length > 0) {
            trader = Trader.fromSaveGame(blocks.get("HANDLER")[0].data, players, islands);
        }

        return new World(
            [...islands.values()],
            [...players.values()],
            task,
            gameName,
            soldiers,
            ships,
            kontors,
            castles,
            cities,
            trader,
        );
    }

    private handleBlock<T>(blocks: Map<string, Block[]>, name: string, entity: any, players: PlayerMap,
                           islands: IslandMap): T[] {
        if (!blocks.has(name)) {
            return [];
        }
        const entities: T[] = [];
        for (const block of blocks.get(name)) {
            while (!block.data.eof()) {
                entities.push(entity.fromSaveGame(block.data, players, islands));
            }
        }
        return entities;
    }

    private async parseIslands(islandBlocks: IslandBlock[], players: PlayerMap) {
        const islands: Map<number, Island> = new Map();
        for (const islandBlock of islandBlocks) {
            const island = Island.fromSaveGame(islandBlock.data);
            const islandBuildingBlocks = islandBlock.inselHausBlocks;
            assert(islandBuildingBlocks.length === 2);
            const islandBottomBlock = islandBuildingBlocks[0];
            const islandTopBlock = islandBuildingBlocks[1];

            if (!island.differsFromBaseIsland) {
                const islandFile = await this.islandLoader.load(island);
                // TODO: Why do we ignore this block
                const islandBasisBlock = this.readBlock(islandFile);
                const islandBaseFieldsBlock = this.readBlock(islandFile);
                // TODO: Are there more blocks?
                island.baseFields = this.parseIslandBuildings(
                    island,
                    islandBaseFieldsBlock,
                    players,
                );
            } else {
                island.baseFields = this.parseIslandBuildings(
                    island,
                    islandBottomBlock,
                    players,
                );
            }

            island.topFields = this.parseIslandBuildings(
                island,
                islandTopBlock,
                players,
            );

            islands.set(island.id, island);
        }
        return islands;
    }

    private parseIslandBuildings(island: Island, block: Block, players: PlayerMap): Array<Array<Field|null>> {
        const data = block.data;
        const dataLength = block.length;

        const fields: Field[][] = [];
        for (let x = 0; x < island.width; x++) {
            fields.push(new Array(island.height).fill(null));
        }

        for (let i = 0; i < dataLength / Field.saveGameDataLength; i++) {
            const field = Field.fromSaveGame(data, players);
            assert(field.x < island.width);
            assert(field.y < island.height);
            fields[field.x][field.y] = field;
        }

        return fields;
    }

    private parsePlayers(block: Block) {
        const players: Map<number, Player> = new Map();
        while (!block.data.eof()) {
            const player = Player.fromSaveGame(block.data);
            players.set(player.id, player);
        }
        return players;
    }

    private readBlock(data: Stream): Block {
        return new Block(data.readString(16), data.read32(), data);
    }
}
