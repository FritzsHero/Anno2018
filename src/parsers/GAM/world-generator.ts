import {islandFromIsland3File, islandFromIsland4File, islandFromIsland5File} from "../../game/world/island";
import World from "../../game/world/world";
import WorldGenerationSettings from "../../game/world/world-generation-settings";
import assert from "../../util/assert";
import {Block} from "./block";
import IslandLoader from "./island-loader";

export default class WorldGenerator {
    constructor(private islandLoader: IslandLoader) {}

    public async populateWorld(
        world: World,
        worldGenerationSettings: WorldGenerationSettings,
    ) {
        let nextIslandId = Math.max(...world.islands.map((island) => island.id)) + 1;

        for (const islandTemplate of worldGenerationSettings.islandTemplates) {
            const {data, isSouth, id: numBaseIsland} = await this.islandLoader.loadRandomIslandFile(
                islandTemplate.size,
                islandTemplate.climate,
            );

            const blocks: Block[] = [];
            while (!data.eof()) {
                blocks.push(Block.fromStream(data));
            }

            const inselBlock = blocks.find((block) => ["INSEL5", "INSEL4", "INSEL3"].includes(block.type));
            assert(inselBlock !== undefined);
            const inselHausBlock = blocks.find((block) => block.type === "INSELHAUS");
            assert(inselHausBlock !== undefined);

            // TODO: Place random things:
            // numNativesNorth,
            // numNativesSouth,
            // numBigIronOre,
            // numSmallIronOre,
            // numGoldOre,
            // numWine,
            // numSugarCane,
            // numSpice,
            // numCacao,
            // numTobacco,
            // numCotton,
            // numTreasures,

            let newIsland;
            switch (inselBlock.type) {
                case "INSEL5":
                    newIsland = islandFromIsland5File(nextIslandId++, islandTemplate.position, isSouth, numBaseIsland,
                        inselBlock.data);
                    break;
                case "INSEL4":
                    newIsland = islandFromIsland4File(nextIslandId++, islandTemplate.position, isSouth, numBaseIsland,
                        inselBlock.data);
                    break;
                case "INSEL3":
                    newIsland = islandFromIsland3File(nextIslandId++, islandTemplate.position, isSouth, numBaseIsland,
                        inselBlock.data);
                    break;
                default:
                    throw new Error("This code cannot be reached.");
            }

            newIsland.baseFields = this.islandLoader.parseIslandBuildings(
                newIsland,
                inselHausBlock,
            );
            newIsland.topFields = this.islandLoader.parseIslandBuildings(
                newIsland,
                Block.empty("INSELHAUS"),
            );

            world.islands.push(newIsland);
        }
    }
}
