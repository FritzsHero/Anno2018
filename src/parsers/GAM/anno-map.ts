import {Island, Ship} from "./gam-parser";

export class AnnoMap {
    constructor(public islands: Island[], public ships: Ship[]) { }
}
