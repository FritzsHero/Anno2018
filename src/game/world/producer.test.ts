import { Block } from "../../parsers/GAM/block";
import { Producer } from "./producer";
import Stream from "../../parsers/stream";

const BLOCK_TYPE = "PRODLIST2";

describe("producer entity", () => {
  it.each([
    new Uint8Array([
      0,
      8,
      14,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0
    ]),
    new Uint8Array([
      0,
      16,
      7,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0
    ])
  ])("loads and saves correctly", (data: Uint8Array) => {
    const DATA_LENGTH = data.length;
    const block = new Block(BLOCK_TYPE, DATA_LENGTH, new Stream(data));

    const producer = Producer.load(block.data);
    expect(producer).toMatchSnapshot();

    const newBlock = new Block(
      BLOCK_TYPE,
      DATA_LENGTH,
      new Stream(new Uint8Array(DATA_LENGTH))
    );
    producer.save(newBlock.data);
    expect(block).toEqual(newBlock);
  });
});
