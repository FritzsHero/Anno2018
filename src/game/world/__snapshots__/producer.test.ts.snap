// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`producer entity loads and saves correctly 1`] = `
Producer {
  "firstGoodStock": 0,
  "goodWasProduced": 0,
  "howMuchIsBeingProduced": 0,
  "islandId": 0,
  "position": Point {
    "x": 8,
    "y": 14,
  },
  "prodCount": 0,
  "secondGoodStock": 0,
  "speedCnt": 8,
  "stock": 0,
  "timeCount": 0,
  "timer": 0,
}
`;

exports[`producer entity loads and saves correctly 2`] = `
Producer {
  "firstGoodStock": 0,
  "goodWasProduced": 0,
  "howMuchIsBeingProduced": 0,
  "islandId": 0,
  "position": Point {
    "x": 16,
    "y": 7,
  },
  "prodCount": 0,
  "secondGoodStock": 0,
  "speedCnt": 8,
  "stock": 0,
  "timeCount": 0,
  "timer": 0,
}
`;
